### ISSUE TITLE
Move to a staging then make it live strategy

### ISSUE THREAD

**Issue by cansavvy** [AlexsLemonade/refinebio-examples#297]
Currently, this repo's github pages are not "live" in that they are not linked to refine.bio and are not directly user-facing. 

However, once we do make this repo's material live #286  we will need to figure out our strategy for editing our material, and it would probably be advisable to have a checkpoint before a PR something becomes user-facing

This issue originated here: https://github.com/AlexsLemonade/refinebio-examples/issues/295#issuecomment-708617271

We will need to figure out the details of what this strategy would look like. It would probably involve a `staging` branch and altering the github actions in a way that we can still see the html previews but they won't be published directly upon merging PRs.

**Comment by cansavvy** [comment]
This is also probably related to #285 and may be something that we need to discuss in tandem.

**Comment by cansavvy** [comment]
I think this will involve a few steps:

1) Make a staging branch and set it so PRs are required. (and whatever other protections we currently have on the master branch should be added here too). 
2) Can we set the default for PRs to go to the `staging` branches (is this an option in settings?). 
3) Update github actions to treat `staging` like the `master` was being treated.

4) Determine processes around deployments:  (aka `staging` -> `master` PRs)
     - How often do we deploy changes?
     - Should we require two approvals for these PRs?

5) CONTRIBUTING.md needs to be updated to reflect this process and any other changes we'll need to make.

**Comment by cansavvy** [comment]
I don't think we want each deployment to have its own version tag number, but we should decide how often we consider something to be a "new release" : https://github.com/AlexsLemonade/refinebio-examples/issues/285

**Comment by jaclyn-taroni** [comment]
> Can we set the default for PRs to go to the `staging` branches (is this an option in settings?).

Yes: https://docs.github.com/en/free-pro-team@latest/github/administering-a-repository/changing-the-default-branch

**Comment by cansavvy** [comment]
As far as this step: 

> 3\. Update github actions to treat `staging` like the `master` was being treated.

I think in most instances we just change the `on` criteria too be `staging` instead of master, but a question is do we want the re-rendering to happen again before `staging -> master` merges or only `some-branch -> staging` merges? I think the answer is run all the checks again before master because why not?

**Comment by jaclyn-taroni** [comment]
> How often do we deploy changes?

This is not like a service where the API might go down and need manual intervention to go back up, so having a set schedule seems a little bit like overkill to me. Which is to say, my answer is as often as we need to.

> Should we require two approvals for these PRs?

Can you say more about why we'd do that? I think the more important question we are really trying to answer is: At what point in the process do we make sure things meet the standards required to be user-facing?

Another question is do we have the notion of [hotfix branches](https://nvie.com/posts/a-successful-git-branching-model/#hotfix-branches)? Say I've found a minor bug or a typo, do we want to go through two rounds of pull requests to fix it?

**Comment by cansavvy** [comment]
> my answer is as often as we need to.

This is probably the answer I expected but wanted to bring it up in case there were other thoughts around it. 

> Can you say more about why we'd do that? I think the more important question we are really trying to answer is: At what point in the process do we make sure things meet the standards required to be user-facing?

I guess just because the stakes are higher because it goes user-facing. However, if the process has worked up until the merge to master, then it shouldn't be necessary. So that's why I'm not completely sold on two approvals, because I could also see that being more annoying than helpful. 

> Another question is do we have the notion of [hotfix branches](https://nvie.com/posts/a-successful-git-branching-model/#hotfix-branches)? Say I've found a minor bug or a typo, do we want to go through two rounds of pull requests to fix it?

I think in the case of the things you've just mentioned, I can see that making sense. I would just be concerned about us using hotfixes too much. But if hotfixes are enabled, then that does answer my question about github actions, aka they should happen every merge for sure.

**Comment by jaclyn-taroni** [comment]
> I guess just because the stakes are higher because it goes user-facing. However, if the process has worked up until the merge to master, then it shouldn't be necessary. So that's why I'm not completely sold on two approvals, because I could also see that being more annoying than helpful.

Agreed. I'm not convinced it would help us. Speaking for myself, I can see getting tired of reviewing something that's going into `staging` and being like "I'll catch it when it goes to `master`" and then not doing that. What we're aiming for with #295 and this issue is that if review has gone on for a while, document what else needs to happen in an issue, merge the outstanding PR, file a new one. This allows you to revisit something with fresh  👀 without scrolling past 10s of comments. 

> I think in the case of the things you've just mentioned, I can see that making sense. I would just be concerned about us using hotfixes too much. But if hotfixes are enabled, then that does answer my question about github actions, aka they should happen every merge for sure.

Yea, I think the policy would be to use hotfixes very judiciously and we would want the same checks for `master` and `staging`. (I see very little downside to checks on both, but if folks have some ideas of what they are please say so!)

**Comment by cansavvy** [comment]
I think the PR process will look like this (correct me where/if I'm wrong). 

PR 1: changes from `some-branch` -> `staging` -> gha re-renders all but does NOT publish to gh-pages. 

PR 2: `staging` -> `master` -> gha re-renders all and publishes to gh-pages. 

Only when changes are straightforward and necessary: 
Hotfix PR: `some-branch` -> `master` -> gha re-renders all and publishes to gh-pages. 

Basically where gha is concerned re-rendering should always happen, but publishing to gh-pages should only happen when merges to master occur.

**Comment by cansavvy** [comment]
`staging` is now the default branch and it has the same protections that `master` has. 

Trying to figure out how to alter gha so publishing only happens when merges to master.

**Comment by cansavvy** [comment]
Something that should be added to #313, but I'm not sure the exact process, is more specifics on how `hotfix` PRs should work. Ultimately we need the hotfixes to make it `master` AND `staging`,  but how do we make sure staging gets them too? Do we file two PRs? How do other repos do this?
